server/src/main/java/org/opensearch/indices/tiering/HotToWarmTieringService.java

```diff
@@ -0,0 +1,875 @@
+     * @param completedIndices set of completed indices
+     * @return true if any of the indices has failed, false otherwise
+     */
+    boolean hasFailedIndices(final Set<IndexTieringInfo> completedIndices) {
+        for (IndexTieringInfo indexTieringInfo : completedIndices) {
+            if (indexTieringInfo.hasFailed()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Updates the request context object in the requestUuidToRequestContext map for
+     * completed (successful/failed) indices.
+     * @param completedIndices set of completed indices
+     */
+    void updateRequestContextForCompletedIndices(final Set<IndexTieringInfo> completedIndices) {
+        for (IndexTieringInfo indexTieringInfo : completedIndices) {
+            String reqId = indexTieringInfo.getRequestUuid();
+            TieringRequestContext requestContext = requestUuidToRequestContext.get(reqId);
+            if (indexTieringInfo.hasFailed()) {
+                requestContext.addToFailed(indexTieringInfo.getIndex(), indexTieringInfo.getFailureReason());
+            } else {
+                requestContext.addToSuccessful(indexTieringInfo.getIndex());
+            }
+        }
+    }
+
+    /**
+     * Updates the cluster state by updating the index metadata for completed (successful/failed) indices.
+     * @param completedIndices set of completed indices
+     */
+    void updateClusterStateForCompletedIndices(final Set<IndexTieringInfo> completedIndices) {
+        clusterService.submitStateUpdateTask(
+            "process hot to warm tiering for completed indices",
+            new ClusterStateUpdateTask(Priority.URGENT) {
+
+                @Override
+                public ClusterState execute(ClusterState currentState) {
+                    final Metadata.Builder metadataBuilder = Metadata.builder(currentState.metadata());
+                    final RoutingTable.Builder routingTableBuilder = RoutingTable.builder(currentState.routingTable());
+                    for (IndexTieringInfo indexTieringInfo : completedIndices) {
+                        final IndexMetadata.Builder indexMetadataBuilder = updateIndexMetadataForCompletedIndex(
+                            metadataBuilder,
+                            indexTieringInfo
+                        );
+                        metadataBuilder.put(indexMetadataBuilder);
+                    }
+                    ClusterState updatedState = ClusterState.builder(currentState)
+                        .metadata(metadataBuilder)
+                        .routingTable(routingTableBuilder.build())
+                        .build();
+
+                    if (hasFailedIndices(completedIndices)) {
+                        // only reroute in case of failed indices to trigger shard relocation for shards to go back to hot nodes
+                        updatedState = allocationService.reroute(updatedState, "hot to warm revert tiering");
```

Comment:
how will this revert ? There is still the index setting for partial data locality on the index ?

Reply:
its updated here: https://github.com/opensearch-project/OpenSearch/pull/14891/files/0813dac038a60a06c433c6a4a65cd0adcbcbe6e8#diff-943058bf06b14acd6356e29328621936b591c5e62481be63b85348892a888233R245